packages/flutter_tools/lib/src/build_system/phases.dart

```diff
@@ -0,0 +1,180 @@
+  /// checking for cycles and naming collisions.
+  ///
+  /// Throws a [StateError] if [phases] is empty.
+  // TODO(jonahwilliams): add output file collision detection.
+  Future<Target> createBuild(Environment environment) async {
+    if (phases.isEmpty) {
+      throw StateError('A BuildDefinition was created with no phases.');
+    }
+    // First create an aggregate target for each phase and store then by.
+    // phase name.
+    final Map<String, _AggregateTarget> aggregates = <String, _AggregateTarget>{};
+    for (BuildPhase buildPhase in phases) {
+      final List<Target> targets = await buildPhase.plan(environment);
+      final _AggregateTarget target = _AggregateTarget(targets, buildPhase);
+      aggregates[target.name] = target;
+    }
+    // Then order the aggregate targets such that dependencies are inserted into
+    // the target graph in order. This is by traversing the list of aggregate
+    // nodes and inserting the dependant aggregate targets into the list of
+    // dependencies. This is validated by the call to [checkCycles] later.
+    final _RootTarget result = _RootTarget(aggregates.values.toList(), this);
+    for (_AggregateTarget target in aggregates.values) {
+      for (String phaseDependency in target.phase.dependencies) {
+        final _AggregateTarget dependency = aggregates[phaseDependency];
+        target.dependencies.add(dependency);
+      }
+    }
+    checkCycles(result);
+    return result;
+  }
+}
+
+// A synthetic target created from a List of [BuildPhase] aggregate targets.
+class _RootTarget extends Target {
+  _RootTarget(this.dependencies, this.definition);
+
+  final BuildDefinition definition;
+
+  @override
+  final List<Target> dependencies;
+
+  @override
+  String get name => definition.name;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  List<Source> get outputs => const <Source>[];
+}
+
+// A synthetic target created from the result of a [BuildPhase.plan].
+class _AggregateTarget extends Target {
+  _AggregateTarget(this.dependencies, this.phase);
+
+  final BuildPhase phase;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  final List<Target> dependencies;
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  String get name => phase.name;
+
+  @override
+  List<Source> get outputs => const <Source>[];
+}
+
+/// A Phase is a creator of [Targets].
+///
+/// A phase doesn't have explicit dependencies on individual targets. Instead,
+/// the build system requires all phases in [dependencies] to have run before
+/// this phase.
+///
+/// A phase will always be run, but the targets that constitute it may be
+/// skipped. The [plan] method of a build's phases are always invoked before
+/// any targets are run.
+abstract class BuildPhase {
```

Comment:
What do you think about a less fancy name, like "TargetGroup" for this? Then you could have "_StaticTargetGroup" and "_DynamicTargetGroup"(s) etc.

Reply:
I'm up for a less fancy name. I don't want to confuse Target + this though, because they do have different functions.